### ISSUE TITLE
NextJS 13: support for server-side components in nextjs-auth0 p2 (app directory is stable now)

### ISSUE THREAD

**Issue by ColemanDunn** [auth0/nextjs-auth0#1193]
### Checklist

- [X] I have looked into the [Readme](https://github.com/auth0/nextjs-auth0#readme), [Examples](https://github.com/auth0/nextjs-auth0/blob/main/EXAMPLES.md), and [FAQ](https://github.com/auth0/nextjs-auth0/blob/main/FAQ.md) and have not found a suitable solution or answer.
- [X] I have looked into the [API documentation](https://auth0.github.io/nextjs-auth0/) and have not found a suitable solution or answer.
- [X] I have searched the [issues](https://github.com/auth0/nextjs-auth0/issues) and have not found a suitable solution or answer.
- [X] I have searched the [Auth0 Community](https://community.auth0.com) forums and have not found a suitable solution or answer.
- [X] I agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Describe the problem you'd like to have solved

Seems like [this issue](https://github.com/auth0/nextjs-auth0/issues/889) can be opened/considered again because as of today app directory is stable: https://vercel.com/home.

Even the link to the beta docs in the [last comment](https://github.com/auth0/nextjs-auth0/issues/889#issuecomment-1503114410) before the issue got locked no longer takes you to the beta docs. Try it yourself: https://beta.nextjs.org/docs

That is all. Thanks!

### Describe the ideal solution

Re-open/consider this issue: https://github.com/auth0/nextjs-auth0/issues/889

**Comment by adamjmcgrath** [comment]
Thanks for raising this @ColemanDunn 

Next.js 13.4 got released 40 mins ago - what took you so long? 😆 

We'll prioritise some work to look into this and update this thread when we have any news.

**Comment by ColemanDunn** [comment]
> Thanks for raising this @ColemanDunn
> 
> Next.js 13.4 got released 40 mins ago - what took you so long? 😆
> 
> We'll prioritise some work to look into this and update this thread when we have any news.

🤣🤣 I happened to get kicked out of the beta docs and redirected to the regular docs as I was looking at them! I remember stumbling across the other issue and seeing your comment only a few hours after you had posted it so I thought I would come back and update. Thanks for the quick response!

**Comment by jorgemurta** [comment]
I'm really looking forward for this! 💪

**Comment by Olivia-li** [comment]
Excited to see this in action :)

**Comment by BRGustavoRibeiro** [comment]
Excited to see this too. I need to code a new project from the ground up and this is dealbreaker to me. :)

**Comment by obouchari** [comment]
Just curious what's the time frame for this to be available in at least a beta version?

**Comment by machulav** [comment]
I just finished my app and got stuck with the final part - auth0 integration with the app directory. So do you have any updates on when this will be ready?

**Comment by glennvgastel** [comment]
For all of you who, like me, are waiting for official support ; I'm currently using this **workaround** that fetches the session on NextJS' server side (`page.tsx` or `layout.tsx`) renders, as well as in any authenticated API calls ;

```typescript
import { headers } from 'next/headers'
import { getSession as auth0GetSession } from '@auth0/nextjs-auth0'
import { IncomingMessage, ServerResponse } from 'http'
import { Socket } from 'net'

const reqRes = () => {
  const req = new IncomingMessage(new Socket())
  headers().forEach((v, k) => {
    req.headers[k] = v
  })
  const res = new ServerResponse(req)
  return { req, res }
}

const getSession = () => {
  const { req, res } = reqRes()
  return auth0GetSession(req, res)
}

export default getSession
```

You can use it like so ;

```typescript
// In `layout.tsx`, for example
const session = await getSession()
if (!session) {
  return redirect('/login')
}
```

The requirement for this to work is, of course, the `appSession` cookie being present. As it's a httpOnly cookie, it should automatically be included with every request.

**Comment by noahebrooks** [comment]
I've been using workarounds for a couple of months now. I recently used Clerk for a new project and realized they have had support for RSC and App dir for a while now.

**Comment by obouchari** [comment]
> I've been using workarounds for a couple of months now. I recently used Clerk for a new project and realized they have had support for RSC and App dir for a while now. 

Same here I like how Clerk not only support edge stuff. But also how they handle the whole Auth flow. But, helas  we're restricted by what our team uses.

**Comment by AzSiAz** [comment]
What's done is done stop complaining and polluting this issue...
Yes it's a shame but the `app directory` was not production ready back then even if marked "stable" (and it should not even be considered production ready right now anyway), even better this workaround has been in the [kitchen-sink](https://github.com/auth0/nextjs-auth0/blob/main/examples/kitchen-sink-example/app/profile-experimental-rsc/page.tsx) exemple for a while now

**Comment by SkyM** [comment]
> stop complaining

If this was a purely open source initiative I would be sympathetic to this mindset, however many of us are paying customers of Auth0 and were sold on the platform based on this type of integration. We have a duty to voice our opinions as customers and let our service provider know how critical this is to our businesses. Auth0 needs to apply more resources to getting this issue addressed in an increasingly competitive environment where other providers like Clerk are ready day one. 

Communication from the Auth0 team simply in terms of a timeline would be greatly appreciated.

**Comment by noahebrooks** [comment]
I wasn't trying to cause the hate train, but it is unfortunate as it seems support for one of the most pressing and groundbreaking features of server components was pushed to the back burner. Especially when we see other SDKs which were super quick moving to add support. I remember when the old thread was still open when 13 was just released the excuse was that the API contract was likely to change many times over and that the feature was too unstable to start a branch which is understandable. However, now its reached stability and there hasn't been any communication in terms of an initiative to support App Dir and RSC. Without communication on this topic I'm hard pressed to keep using Auth0 as I would like my company to be able to take advantage of RSC. I would be more than happy to help.

**Comment by buckion** [comment]
Part of the sales value proposition of Auth0 is the client libraries. If I'm needing to allocate team time to write them, then I'm paying for an inconvenient and slow database. This is making me quite reticent to expand my contract. Are people in this thread aware of good options that support data residency in multiple countries and SSO connections? 

It's seeming like Supabase might actually meet most/all of those needs as of very recently, and it has functioning libraries.

**Comment by SkyM** [comment]
@buckion We have been evaluating Clerk, which while new has a lot of attractive features. The blocker for us is the lack of cross-application user support. I would also be interested in anybody voicing alternatives to Auth0.

**Comment by AzSiAz** [comment]
Sorry for the misunderstanding but don't get me wrong: you should complain
Especially if you are a paying customer, just not here do it by others, and more appropriate, channels which you certainly have since as you said you are a paying customer, it's the only way they will move forward faster. For all we know no product manager is even watching this GitHub repository and/or no one report back on what happen here. And given the lack of pull request NextJS doesn't even seem to be a priority...

So yes I stand by what I said, stop complaining here, it's just noise for people watching this issue to know when it's done or for people who could be replying to ask something more than just basic support for server-side components
There is a Auth0 community forum [here](https://community.auth0.com/) if you want to keep it public

But I do agree that a timeline would be most welcome to know when we can start playing with it or to plan it's implementation on our side, even if it's behind a `/beta` import for a little while 

Good luck with all your project everyone and let's hope we can get it soon officially 😄

**Comment by noahebrooks** [comment]
@buckion I agree with @SkyM, Clerk has proven to be a notable alternative. I was working on a side project over the weekend and was able to implement auth with the App dir in under 5 minutes using Clerk, I was completely blown away. We use Supabase as a company and their recent SSO connections prove to be promising as an overall backend Auth/DB solution, also direct RLS makes Supabase IMO possibly the best provider for small teams. Note Clerk does have a Supabase plugin to enable RLS on Supabase using Clerk as the Identity Provider. Clerk comes with a number of react components which makes creating login/signup/profile/... crazy simple which is a huge plus, as well as components like a profile switcher and org switcher which is great for applications in which the user can have multiple workspaces/orgs. We used to use Logto-io as a self hosted solution, had a few problems but not terrible either.

**Comment by ColemanDunn** [comment]
> Yes it's a shame but the `app directory` was not production ready back then even if marked "stable" (and it should not even be considered production ready right now anyway)

Where are you getting this information? App directory was never "marked stable" until it was next 13.4 (no experimental flag needed in next.config.js). Not is it only is it production ready but it is _reccomended_ by `npx create-next-app`

**Comment by catchshyam** [comment]
I am sorry but what we really need is a beta release date. I wish I could tell my product manager that I can comit for our product release shortly but he is asking me for a date which is dependent on Auth0's release date. Guys, please understand, this is a paid product not an opensource initiative.

**Comment by catchshyam** [comment]
In addition to the work around mentioned above to verify the user session on the pages, do anyone know how we can get the access token that I want to forward to my API server? I tried to extend the workaround to get the access token like below but the JWT is malformed

**Comment by noahebrooks** [comment]
@catchshyam I found this to be a suitable workaround atm to get the user's access token.

```javascript
import * as jwt from 'jsonwebtoken';
import jwksClient from 'jwks-rsa';

export async function verifyAccessTokenFromRequest(req: Request) {
  const authHeader = req.headers.get('authorization');
  if (!authHeader) {
    return null;
  }

  const token = authHeader.substring(7);
  const endpoint = process.env.AUTH0_JWKS_ENDPOINT || '';
  const client = jwksClient({
    jwksUri: endpoint,
  });

  const kid = process.env.AUTH0_JWKS_KEY_ID || '';
  const signingKey = await client.getSigningKey(kid);

  try {
    return jwt.verify(token, signingKey.getPublicKey());
  } catch(error) {
    return null;
  }
}
```

**Comment by noahebrooks** [comment]
Misread the question that is for getting and verifying the access token server side^

**Comment by macrigiuseppe** [comment]
@catchshyam once you have the `session` object you can simply access the token by simply doing `session.accessToken`

**Comment by adamjmcgrath** [comment]
Hi all 👋 just to let you know that we're still actively working on this feature - you can check out the branch here https://github.com/auth0/nextjs-auth0/tree/server-components (including an example app here https://github.com/auth0/nextjs-auth0/tree/server-components/example-app/app)

We're almost feature complete, but we still have a few things to figure out - so can't give a specific date for when we might make it available as a beta. But we hope this will be in the next week or two.

**Comment by buckion** [comment]
Hey @adamjmcgrath , thankyou for the update! 

What are some of the things to be figured out? Can people join in on the effort to help push it over the line?

**Comment by noahebrooks** [comment]
@adamjmcgrath I second this^

**Comment by talboren** [comment]
Any news here? 👀

**Comment by buckion** [comment]
Update: 3.0.0 beta is live on npm!

**Comment by adamjmcgrath** [comment]
Yep, v3.0.0-beta.0 is available on npm.

For more information on the Beta, visit https://github.com/auth0/nextjs-auth0/issues/1235

Closing in favour of https://github.com/auth0/nextjs-auth0/issues/1235

**Comment by ColemanDunn** [comment]
Does anyone have any info on how to refresh the session.user from useUser()? When I update a user's role (from the api or the dashboard, wherever) it is not reflected in the session.user object. We used to have a custom handler in handleAuth (with the old req and res) that would do it but now I can't seem to get it working with app directory. I feel like there has got to be an easier way to simply update the user object than making a custom handler. Anyone have some insight?
